### PR TITLE
Add computed fields to SQL

### PR DESCRIPTION
**Pull request by mohamedsalem401**

### CODE REVIEW

File: src/lib/schema.ts [hunk 1]
```diff
@@ -25,50 +25,61 @@ interface File {
...
-      table.string("filetype"); // Type field in frontmatter if it exists
-      table.string("metadata"); // All frontmatter data
+      table.string("file_path").unique().notNullable();
+      table.string("extension").notNullable();
+      table.string("url_path");
+      table.string("filetype");
+      table.string("metadata");
+      for (let index = 0; index < properties.length; index++) {
+        table.string(properties[index]);
+      }
```
Comment (olayway) [inline src/lib/schema.ts:82]: What happens if a computed property returns an array or a boolean? Shouldn't it be `table.<some type>`?  Also, were the comments incorrect?

```suggestion
      properties.forEach((property) => {
        table.string(property);
      })
```

File: src/lib/schema.ts [hunk 2]
```diff
@@ -88,15 +99,58 @@ class MddbFile {
...
+      const defaultProperties = [
+        "_id",
+        "file_path",
+        "extension",
+        "url_path",
+        "filetype",
+      ];
+
+      for (const key in file) {
+        if (Object.prototype.hasOwnProperty.call(file, key)) {
```
Comment (olayway) [inline src/lib/schema.ts:114]: this will always resolve to `true`

File: src/lib/schema.ts [hunk 3]
```diff
@@ -88,15 +99,58 @@ class MddbFile {
...
+          const value = file[key];
+          // If the value is undefined, default it to null
+          serializedFile[key] = defaultProperties.includes(key)
+            ? value
+            : value !== undefined
+            ? JSON.stringify(value)
+            : null;
+        }
```
Comment (olayway) [inline src/lib/schema.ts:122]: The comment is misleading. What are you trying to achieve here? There must be a better way to code this...

File: src/lib/markdowndb.ts [hunk 4]
```diff
@@ -80,14 +82,36 @@ export class MarkdownDB {
...
+      "metadata",
+      "tags",
+      "links",
+    ];
+    await MddbFile.createTable(
+      this.db,
+      properties.filter(
+        (property) => defaultProperties.indexOf(property) === -1
+      )
+    );
```
Comment (olayway) [inline src/lib/markdowndb.ts:113]: Wait, we hadn't been creating the files table before at all?!

File: src/lib/markdowndb.ts [hunk 5]
```diff
@@ -80,14 +82,36 @@ export class MarkdownDB {
...
+    const defaultProperties = [
+      "_id",
+      "file_path",
+      "extension",
+      "url_path",
+      "filetype",
+      "metadata",
+      "tags",
+      "links",
+    ];
```
Comment (olayway) [inline src/lib/markdowndb.ts:107]: I'd save this information on the MddbFile class as a static prop.

File: src/lib/markdowndb.ts [hunk 6]
```diff
@@ -80,14 +82,36 @@ export class MarkdownDB {
...
+    await MddbFile.createTable(
+      this.db,
+      properties.filter(
+        (property) => defaultProperties.indexOf(property) === -1
+      )
```
Comment (olayway) [inline src/lib/markdowndb.ts:112]: It doesn't look good, move any detailed logic related to table creation the createTable method.

File: src/lib/schema.ts [hunk 7]
```diff
@@ -25,50 +25,61 @@ interface File {
   url_path: string | null;
   filetype: string | null;
   metadata: MetaData | null;
+  [key: string]: string | null | MetaData | undefined;
```
Comment (olayway) [inline src/lib/schema.ts:28]: Rather:

```suggestion
  [key: string]: unknown;
```

**Review by olayway — CHANGES_REQUESTED** [review]
Good start but can be improved ;)

File: src/lib/schema.ts [hunk 8]
```diff
@@ -88,15 +99,58 @@ class MddbFile {
...
+    return this.file.url_path;
+  }
+
+  get filetype(): string | null {
+    return this.file.filetype;
+  }
+
+  get metadata(): MetaData | null {
+    return this.file.metadata;
+  }
```
[the remark this answers is not in the document]
Comment (mohamedsalem401) [reply src/lib/schema.ts:153]: We need this!

File: src/lib/schema.ts [hunk 9]
```diff
@@ -25,50 +25,61 @@ interface File {
...
+        key !== "_id" &&
+        key !== "file_path" &&
+        key !== "extension" &&
+        key !== "url_path" &&
+        key !== "filetype" &&
+        key !== "metadata"
+      ) {
+        this.file[key] = file[key];
+      }
+    }
```
[the remark this answers is not in the document]
Comment (mohamedsalem401) [reply src/lib/schema.ts:65]: I agree

File: src/lib/schema.ts (hunk 2 above)
Comment (mohamedsalem401) [reply src/lib/schema.ts:114]: Will fix this

File: src/lib/schema.ts [hunk 10]
```diff
@@ -25,50 +25,61 @@ interface File {
...
-  filetype: string | null;
-  metadata: MetaData | null;
+  file: File = {
+    _id: "",
+    file_path: "",
+    extension: "",
+    url_path: null,
+    filetype: null,
+    metadata: null,
+  };
```
[the remark this answers is not in the document]
Comment (mohamedsalem401) [reply src/lib/schema.ts:42]: But this way we lose tract of all fields, it's like using `any` as a type. resulting in a loss of type safety.

File: src/lib/schema.ts (hunk 1 above)
Comment (mohamedsalem401) [reply src/lib/schema.ts:82]: I see what you are saying, I thought about a couple of scenarios:
- a field is array => there's no way to store arrays in SQL
- a field is a number => what type of number? float, double, ...
- what if a field is an object => there's no way to store objects in SQL
- what if a field has a different type ( number/string ) across two files, should it be defaulted to string?

File: src/lib/markdowndb.ts (hunk 4 above)
Comment (mohamedsalem401) [reply src/lib/markdowndb.ts:113]: We did but was in `resetDatabase` function, but now it has a different API ( needs to receive a list of properties to include as columns )

File: src/lib/schema.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (olayway) [reply src/lib/schema.ts:42]: That't not what I meant. Use original properties + this one for dynamic ones (aka computed fields)

File: src/lib/schema.ts (hunk 1 above)
Comment (olayway) [reply src/lib/schema.ts:82]: Well, use any SQL types that are most suited for each JS/TS type. But it's not like everything needs to be stringified 😅 If it's an object/array, stringify to JSON. 

> what if a field has a different type ( number/string ) across two files, should it be defaulted to string?
You could type the computed function signature in a way that it always should return a single type defined by the user. But for now, ok, just stringify everything. We'll properly implement it when we move computed fields definitions to their final destination = document types.

**Pull request merged**